ikra/soa/array_operator.h

```diff
@@ -0,0 +1,260 @@
+#ifndef SOA_ARRAY_OPERATOR_H
+#define SOA_ARRAY_OPERATOR_H
+
+#include "soa/array_field.h"
+#include "soa/util.h"
+
+// This is used for operations between std::arrays and array fields
+#define IKRA_TEMPLATE_ARRAY_FIELD \
+  template<typename T, size_t ArraySize, \
+           IndexType Capacity, uint32_t Offset, \
+           int AddressMode, int StorageMode, typename Owner>
+// This is used for operations between two array fields
+#define IKRA_TEMPLATE_ARRAY_FIELD_W \
+  template<typename T, size_t ArraySize, \
+           IndexType Capacity1, uint32_t Offset1, \
+           int AddressMode1, int StorageMode1, typename Owner1, \
+           IndexType Capacity2, uint32_t Offset2, \
+           int AddressMode2, int StorageMode2, typename Owner2>
+
+// For `IKRA_TEMPLATE_ARRAY_FIELD`, just write `XXX_ARRAY_FIELD()`;
+// For `IKRA_TEMPLATE_ARRAY_FIELD_W`, write like `XXX_ARRAY_FIELD(1)`.
+#define IKRA_AOS_ARRAY_FIELD(n) AosArrayField_<std::array<T, ArraySize>, \
+  Capacity ## n, Offset ## n, AddressMode ## n, StorageMode ## n, Owner ## n>
+#define IKRA_SOA_ARRAY_FIELD(n) SoaArrayField_<T, ArraySize, \
+  Capacity ## n, Offset ## n, AddressMode ## n, StorageMode ## n, Owner ## n>
+#define IKRA_STD_ARRAY std::array<T, ArraySize>
+
+// TODO: Provide overload where `left` and `right` are passed as reference
+#define IKRA_DEFINE_IKRA_STD_ARRAY_OPERATORS_VECTOR(symbol) \
```

Comment:
Comment like "Given a symbol, defines an operator between all combinations of ...".